### PR TITLE
Refactor Dataloading and Dataset

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*
#755 
#756 

*Description of changes:*
Change the Dataloader API and Dataset API according to #755 and #756.
Update the unit tests accordingly
Update the training and inference codes accordingly.

TODO: Update examples
TODO: Update docs.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/dataloading/dataset.py [hunk 1]
```diff
@@ -354,780 +389,511 @@ def get_edge_feats(self, input_edges, edge_feat_field, device='cpu'):
...
+        g = self._g
+        pb = g.get_partition_book()
+        train_idxs = {}
+        num_train = 0
+        ntypes = self._check_ntypes(ntypes)
+        masks = self._check_node_mask(ntypes, mask)
 
-    def prepare_data(self, g):
-        """ Prepare the testing edge set if any
+        for ntype, msk in zip(ntypes, masks):
```
Comment (zhjwy9343) [inline python/graphstorm/dataloading/dataset.py:512]: ditto for `msk`

File: python/graphstorm/dataloading/dataset.py [hunk 2]
```diff
@@ -354,780 +389,511 @@ def get_edge_feats(self, input_edges, edge_feat_field, device='cpu'):
...
-                infer_idx = dgl.distributed.edge_split(
-                    th.full((g.num_edges(canonical_etype),), True, dtype=th.bool),
-                    pb, etype=canonical_etype, force_even=True)
-                infer_idxs[canonical_etype] = infer_idx
+        idxs = {}
+        num_data = 0
+        ntypes = self._check_ntypes(ntypes)
+        masks = self._check_node_mask(ntypes, mask)
+
+        if isinstance(mask, str):
```
Comment (zhjwy9343) [inline python/graphstorm/dataloading/dataset.py:543]: Should the `self._check_node_mask()` do the same thing for the line #543 to #551?

File: python/graphstorm/dataloading/dataset.py [hunk 3]
```diff
@@ -354,780 +389,511 @@ def get_edge_feats(self, input_edges, edge_feat_field, device='cpu'):
...
+            # mask is a string
+            # validation/test masks are using the same name
+            masks = [mask] * len(ntypes)
+
+        assert len(ntypes) == len(masks), \
+            "Expecting the number of ntypes matches the number of mask fields, " \
+            f"But get {len(ntypes)} and {len(masks)}." \
+            f"The node types are {ntypes} and the mask fileds are {masks}"
+        for ntype, msk in zip(ntypes, masks):
```
Comment (zhjwy9343) [inline python/graphstorm/dataloading/dataset.py:552]: ditto for the `msg`

File: python/graphstorm/dataloading/dataset.py [hunk 4]
```diff
@@ -354,780 +389,511 @@ def get_edge_feats(self, input_edges, edge_feat_field, device='cpu'):
...
+        mask: str or list of str
+            The node feature field storing the training mask.
+            Default: "train_mask"
 
-    @property
-    def pos_graph_feat_field(self):
-        """ Get edge feature fields of positive graphs
+        Returns
+        -------
+        dict of Tensors : The returned nodes
```
Comment (zhjwy9343) [inline python/graphstorm/dataloading/dataset.py:466]: "returned nodes" -> "returned node indexes"?

File: python/graphstorm/dataloading/dataset.py [hunk 5]
```diff
@@ -354,780 +389,511 @@ def get_edge_feats(self, input_edges, edge_feat_field, device='cpu'):
...
-            # DGL Graph edge type is not canonical. It is just list[str].
-            assert self._g.ntypes == [DEFAULT_NTYPE] and \
-                   self._g.etypes == [DEFAULT_ETYPE[1]], \
-                f"It is required to be a homogeneous graph when target_etype is not provided " \
-                f"or is set to {DEFAULT_ETYPE} on edge tasks, expect node type " \
-                f"to be {[DEFAULT_NTYPE]} and edge type to be {[DEFAULT_ETYPE[1]]}, " \
-                f"but get {self._g.ntypes} and {self._g.etypes}"
+        Returns
+        -------
+        dict of Tensors : The returned training masks
```
Comment (zhjwy9343) [inline python/graphstorm/dataloading/dataset.py:503]: "training masks" -> "training node indexes"?

File: python/graphstorm/dataloading/dataset.py [hunk 6]
```diff
@@ -354,780 +389,511 @@ def get_edge_feats(self, input_edges, edge_feat_field, device='cpu'):
...
+        mask: str or list of str
+            The node feature field storing the validation mask.
+            Default: "val_mask"
 
-#### Node classification/regression Task Data ####
-class GSgnnNodeData(GSgnnData):  # pylint: disable=abstract-method
-    """ Data for node tasks
+        Returns
+        -------
+        dict of Tensors : The returned validation masks
```
Comment (zhjwy9343) [inline python/graphstorm/dataloading/dataset.py:580]: "validation masks" -> "validation node indexes"?

File: python/graphstorm/dataloading/dataset.py [hunk 7]
```diff
@@ -354,780 +389,511 @@ def get_edge_feats(self, input_edges, edge_feat_field, device='cpu'):
...
+        ntypes: str or list of str
+            Node types to get the test set.
+        mask: str or list of str
+            The node feature field storing the test mask.
+            Default: "test_mask"
 
         Returns
         -------
-        dict of Tensors : the returned node labels.
+        dict of Tensors : The returned test masks
```
Comment (zhjwy9343) [inline python/graphstorm/dataloading/dataset.py:600]: "test masks" -> "test node indexes"?

**Review by zhjwy9343 — COMMENTED** [review]
Will continue reviewing.

File: python/graphstorm/dataloading/utils.py [hunk 8]
```diff
@@ -136,3 +136,44 @@ def flip_node_mask(dist_tensor, indices):
...
+
+def verify_label_field(label_field):
+    """ Verify the format of label fields
+
+        Parameters
+        ----------
+        label_field: str or dict of str
+    """
+    assert label_field is not None and \
+        (isinstance(label_field, str) or \
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/dataloading/utils.py:148]: No.

File: python/graphstorm/dataloading/utils.py [hunk 9]
```diff
@@ -136,3 +136,44 @@ def flip_node_mask(dist_tensor, indices):
...
+        Parameters
+        ----------
+        node_feats: str, or dist of list of str
+            str: All the nodes have the same feature name.
+            list of string: All the nodes have the same list of features.
+            dist of list of string: Each node type have different set of node features.
+    """
+    assert node_feats is None or \
+            isinstance(node_feats, str) or \
+            (isinstance(node_feats, dict) and \
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/dataloading/utils.py:166]: It can be provided as `_N: [xxx, xxx]`

[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/dataloading/utils.py:166]: Currently, the GraphStorm argparser does not support list of string.
We can open an issue for that.

File: python/graphstorm/dataloading/utils.py [hunk 10]
```diff
@@ -136,3 +136,44 @@ def flip_node_mask(dist_tensor, indices):
...
+                isinstance(list(node_feats.values())[0], list) and \
+                isinstance(list(node_feats.values())[0][0], str)), \
+                "Node features must be a string, " \
+                f"or a dict of list of string, but get {node_feats}."
+
+def verify_edge_feat_fields(edge_feats):
+    assert edge_feats is None or \
+            isinstance(edge_feats, str) or \
+            (isinstance(edge_feats, dict) and \
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/dataloading/utils.py:184]: Same as the node features.

File: python/graphstorm/dataloading/dataset.py [hunk 11]
```diff
@@ -76,13 +75,27 @@ def prepare_batch_input(g, input_nodes,
...
                     data = data.gather(nid.to(dev))
                 else:
                     data = data[nid].to(dev)
                 feats.append(data)
-            feat[ntype] = th.cat(feats, dim=1)
+            assert len(feats) > 0, \
+                "No feature exists in the graph. " \
+                f"Expecting the graph have following node features {feat_name}."
+
+            if len(feats[0].shape) == 1:
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/dataloading/dataset.py:90]: For 1D features, it should be stored as `(num_nodes, 1)` instead of `(num_nodes,)`.
GConstruct and GSProcessing will automatically convert 1D features into 2D like https://github.com/awslabs/graphstorm/blob/main/python/graphstorm/gconstruct/transform.py#L1310-L1316

File: python/graphstorm/dataloading/dataset.py [hunk 12]
```diff
@@ -108,22 +121,37 @@ def prepare_batch_edge_input(g, input_edges,
...
                 else:
                     data = data[eid].to(dev)
                 feats.append(data)
-            feat[etypes] = th.cat(feats, dim=1)
+
+            assert len(feats) > 0, \
+                "No feature exists in the graph. " \
+                f"Expecting the graph have following edge features {feat_name}."
+
+            if len(feats[0].shape) == 1:
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/dataloading/dataset.py:146]: See above

File: python/graphstorm/dataloading/dataset.py [hunk 13]
```diff
@@ -138,29 +166,37 @@ class GSgnnData():
...
+        Default: None
     """
 
-    def __init__(self, graph_name, part_config, node_feat_field, edge_feat_field,
-                 decoder_edge_feat=None, lm_feat_ntypes=None, lm_feat_etypes=None):
+    def __init__(self, part_config, node_feat_field=None, edge_feat_field=None,
+                 lm_feat_ntypes=None, lm_feat_etypes=None):
+        graph_name = get_graph_name(part_config)
         self._g = dgl.distributed.DistGraph(graph_name, part_config=part_config)
+        self._graph_name = graph_name
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/dataloading/dataset.py:186]: Either way is fine. I would prefer the current impl.

File: python/graphstorm/dataloading/dataset.py [hunk 14]
```diff
@@ -354,780 +389,511 @@ def get_edge_feats(self, input_edges, edge_feat_field, device='cpu'):
...
-    r""" Edge prediction inference data
+        g = self.g
+        pb = g.get_partition_book()
+        unlabeled_idxs = {}
+        num_unlabeled = 0
+        ntypes = list(train_idxs.keys())
+        masks = self._check_node_mask(ntypes, mask)
 
-    GSgnnEdgeInferData prepares the data for edge prediction inference.
+        for ntype, msk in zip(ntypes, masks):
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/dataloading/dataset.py:475]: `mask` is the name of an input parameter. Using `msk` here can avoid potential bugs.

File: python/graphstorm/dataloading/dataset.py (hunk 2 above)
Comment (classicsong) [reply python/graphstorm/dataloading/dataset.py:543]: Good catch

File: python/graphstorm/dataloading/dataset.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/dataloading/dataset.py:90]: There is one case where users can still use DGL graph as the input of graph partition. So we might not be able to control the dimension of features. In such case, it could be better to change this `if len...` to an assertion to let users know they must store feature in `(num_nodes, 1)` format.

[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/dataloading/dataset.py:90]: Update: https://graphstorm.readthedocs.io/en/latest/tutorials/own-data.html#option-2-required-dgl-graph

[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/dataloading/dataset.py:90]: All nodes/edges features must be 2D tensors.

File: python/graphstorm/dataloading/dataloading.py [hunk 15]
```diff
@@ -1271,13 +1542,20 @@ class GSgnnNodeDataLoader(GSgnnNodeDataLoaderBase):
         from graphstorm.trainer import GSgnnNodePredictionTrainer
 
         np_data = GSgnnNodeTrainData(...)
-        np_dataloader = GSgnnNodeDataLoader(np_data, target_idx, fanout=[15, 10], batch_size=128)
+        np_dataloader = GSgnnNodeDataLoader(np_data, target_idx, fanout=[15, 10],
```
Comment (zhjwy9343) [inline python/graphstorm/dataloading/dataloading.py:1545]: The above line still use GSgnnNodeTrainerData. Might change to GSgnnData.

**Review by zhjwy9343 — COMMENTED** [review]
Will continue to review models, run, and tests.

File: python/graphstorm/run/gsgnn_np/gsgnn_np.py [hunk 16]
```diff
@@ -85,10 +81,17 @@ def main(config_args):
         trainer.restore_model(model_path=config.restore_model_path,
                               model_layer_to_load=config.restore_model_layers)
     trainer.setup_device(device=get_device())
+    train_idxs = train_data.get_node_train_set(config.target_ntype, mask="train_mask")
```
Comment (zhjwy9343) [inline python/graphstorm/run/gsgnn_np/gsgnn_np.py:84]: As by default the `get_node_train_set` uses "train_mask" as mask name, we can remove the mask argument here to simplify our own code.

File: python/graphstorm/run/gsgnn_np/gsgnn_np.py [hunk 17]
```diff
@@ -85,10 +81,17 @@ def main(config_args):
...
+
+    eval_ntype = config.eval_target_ntype \
+        if config.eval_target_ntype is not None else config.target_ntype
+    val_idxs = train_data.get_node_val_set(eval_ntype, mask="val_mask")
```
Comment (zhjwy9343) [inline python/graphstorm/run/gsgnn_np/gsgnn_np.py:88]: Same, may remove the `mask="val_mask"`.

File: python/graphstorm/run/gsgnn_np/gsgnn_np.py [hunk 18]
```diff
@@ -85,10 +81,17 @@ def main(config_args):
...
+    test_idxs = train_data.get_node_test_set(eval_ntype, mask="test_mask")
```
Comment (zhjwy9343) [inline python/graphstorm/run/gsgnn_np/gsgnn_np.py:89]: Same, may remove the `mask="test_mask"`.

File: python/graphstorm/run/gsgnn_np/gsgnn_np.py [hunk 19]
```diff
@@ -98,34 +101,44 @@ def main(config_args):
 
     if config.use_pseudolabel:
         # Use nodes not in train_idxs as unlabeled node sets
-        unlabeled_idxs = train_data.get_unlabeled_idxs()
+        unlabeled_idxs = train_data.get_unlabeled_node_set(train_idxs, mask="train_mask")
```
Comment (zhjwy9343) [inline python/graphstorm/run/gsgnn_np/gsgnn_np.py:104]: Same, may remove the `mask="train_mask"`.

File: python/graphstorm/trainer/glem_np_trainer.py [hunk 20]
```diff
@@ -66,7 +66,8 @@ class GLEMNodePredictionTrainer(GSgnnNodePredictionTrainer):
             "my_graph", "/path/to/part_config", "my_node_type")
         target_idx = {"my_node_type": target_nodes_tensor}
         my_data_loader = GSgnnNodeDataLoader(
-            my_dataset, target_idx, fanout=[10], batch_size=1024, device='cpu')
+            my_dataset, target_idx, fanout=[10], batch_size=1024,
```
Comment (zhjwy9343) [inline python/graphstorm/trainer/glem_np_trainer.py:69]: Need to change the `GSgnnNodeTrainData` to `GSgnnData` in the line #61.

File: python/graphstorm/trainer/np_trainer.py [hunk 21]
```diff
@@ -62,7 +62,8 @@ class GSgnnNodePredictionTrainer(GSgnnTrainer):
             "my_graph", "/path/to/part_config", "my_node_type")
         target_idx = {"my_node_type": target_nodes_tensor}
         my_data_loader = GSgnnNodeDataLoader(
-            my_dataset, target_idx, fanout=[10], batch_size=1024)
+            my_dataset, target_idx, fanout=[10], batch_size=1024,
```
Comment (zhjwy9343) [inline python/graphstorm/trainer/np_trainer.py:65]: Need to change the line #61. "GSgnnNodeTrainData" to "GSgnnData".

**Comment by zhjwy9343** [comment]
One major concern of the new `GSgnnData` is about how it provides the metadata to avoid conflicts with the given arguments, e.g., `node_feat_field` and `edge_feat_field`. 

For example, if users want to extract all node feature fields of a graph data, will the data.node_feat_field() method return a dictionary of all nodes with feature fields? The current implementation will return "None" is not set this argument in the constructor, or the given values. This may cause confusion.

So can we have a set of dedicate methods in the `GSgnnData` to return the metadata of a graph dataset, e.g., `get_all_node_feat_fields()`, `get_all_edge_feat_fields()`, and etc.

File: python/graphstorm/dataloading/dataset.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/dataloading/dataset.py:90]: Updated

**Comment by classicsong** [comment]
> One major concern of the new `GSgnnData` is about how it provides the metadata to avoid conflicts with the given arguments, e.g., `node_feat_field` and `edge_feat_field`.
> 
> For example, if users want to extract all node feature fields of a graph data, will the data.node_feat_field() method return a dictionary of all nodes with feature fields? The current implementation will return "None" is not set this argument in the constructor, or the given values. This may cause confusion.
> 
> So can we have a set of dedicate methods in the `GSgnnData` to return the metadata of a graph dataset, e.g., `get_all_node_feat_fields()`, `get_all_edge_feat_fields()`, and etc.

Even with get_all_node_feat_fields and get_all_edge_feat_fields does not solve the problem. The key problem is because data.node_feat_field is from the argument node_feat_field.

**Comment by zhjwy9343** [comment]
> > One major concern of the new `GSgnnData` is about how it provides the metadata to avoid conflicts with the given arguments, e.g., `node_feat_field` and `edge_feat_field`.
> > For example, if users want to extract all node feature fields of a graph data, will the data.node_feat_field() method return a dictionary of all nodes with feature fields? The current implementation will return "None" is not set this argument in the constructor, or the given values. This may cause confusion.
> > So can we have a set of dedicate methods in the `GSgnnData` to return the metadata of a graph dataset, e.g., `get_all_node_feat_fields()`, `get_all_edge_feat_fields()`, and etc.
> 
> Even with get_all_node_feat_fields and get_all_edge_feat_fields does not solve the problem. The key problem is because data.node_feat_field is from the argument node_feat_field.

As we discuss, we need to have a `get_graph_metadata()` method to return all information of a given graph.

**Review by zhjwy9343 — APPROVED** [review]
LGTM
